proposals/ipfs-pubsub-content-routing-v1.md

```diff
@@ -0,0 +1,156 @@
+-->
+
+#### Assumptions &amp; hypotheses
+_What must be true for this project to matter?_
+<!--(bullet list)-->
+- There are users who currently have too much data to provide in the IPFS DHT
+- Users who cannot provide all their data in the IPFS DHT have difficulty making their data available to the people who want to download that data
+- Having a non-optimal short term solution to this problem is better than waiting for a longer term solution
+
+#### User workflow example
+_How would a developer or user use this new capability?_
+<!--(short paragraph)-->
+
+User with many objects (hub - name TBD): They add a config option in go-ipfs indicating that they want to be a hub.
+
+All users: By default they automatically send queries to the hub pubsub channel in addition to the DHT, but it can be disabled.
+
+#### Impact
+_How would this directly contribute to web3 dev stack product-market fit?_
+
+It supports groups with large amount of data that want to make their data available via the stack. This is particularly useful since users who don't want to run their own infrastructure will get some larger bodies (e.g. centralized pinning services, Filecoin miners, etc.) to store a copy of their data for them.
+
+<!--
+Explain how this addresses known challenges or opportunities.
+What awesome potential impact/outcomes/results will we see if we nail this project?
+-->
+
+#### Leverage
+_How much would nailing this project improve our knowledge and ability to execute future projects?_
+
+<!--
+Explain the opportunity or leverage point for our subsequent velocity/impact (e.g. by speeding up development, enabling more contributors, etc)
+-->
+
+Solving this problem should increase the number of projects and collaborations in our ecosystem as well as provide information on how people are using content routing that could be useful in future designs.
+
+#### Confidence
+_How sure are we that this impact would be realized? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+<!--Explain why this rating-->
+3 . Among users with a large amount of data figuring out how to make that data findable is a top request.
+
+## Project definition
+#### Brief plan of attack
+
+<!--Briefly describe the milestones/steps/work needed for this project-->
+
+The idea is that hubs will each subscribe to a pubsub channel (e.g. `/ipfs/content-routing/1.0.0`) and when users are looking for who has content they'll publish to that pubsub channel a request for the CID they are looking for. When a hub receives an inbound request for a CID they will check if they have it and if so they'll connect back to the user who will then end up fetching the data via Bitswap.

```

Comment:
If all of the items below were implemented, then we could also add another layer on top of generic content-routing for specific topic content routing, correct? So e.g. a video sharing service could sit on top of the general IPFS implementation and use these same mechanisms for finding CIDs but not crowd out the general space, `/ipfs/vidshare-contentrouting/1.0.0`.
So maybe this proposal is about the primitives for content routing, with a general implementation but also the possibility of focused topic use-cases too? A dapp could then just use all of these primitives for discovering CID providers and fetching content but be subscribed to an alternative broadcast channel.

Reply:
Yep, for efficiency we'd probably want to modify pubsub to be able to simultaneous broadcast messages across multiple topics without repeating itself. It's part of the spec and something we previously had pseudo-implemented, but if there's actual value there then it's probably worth implementing.